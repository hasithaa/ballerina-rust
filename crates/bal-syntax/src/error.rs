use std::fmt;

#[derive(Debug)]
pub enum LexerError {
    InvalidCharacter {
        character: char,
        position: usize,
        line: usize,
        column: usize,
        line_content: String,
    },
    UnterminatedString(usize),
    UnexpectedEof,
}

#[derive(Debug)]
pub enum ParserError {
    UnexpectedToken {
        expected: Vec<String>,
        found: String,
        span: Span,
    },
    MissingToken {
        expected: String,
        after: String,
        span: Span,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    pub file: Option<String>,
    pub start: usize,
    pub end: usize,
    pub line: usize,
    pub column: usize,
    pub line_content: Option<String>,
}

impl fmt::Display for LexerError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LexerError::InvalidCharacter { character, line, column, line_content, .. } => {
                writeln!(f, "error: invalid character")?;
                writeln!(f, " --> line {}:{}", line, column)?;
                writeln!(f, "  |")?;
                writeln!(f, "{} | {}", line, line_content)?;
                writeln!(f, "  | {}^ unexpected character '{}'",
                    " ".repeat(*column),
                    if character.is_control() {
                        format!("\\x{:02x}", *character as u8)
                    } else {
                        character.to_string()
                    }
                )
            }
            LexerError::UnterminatedString(pos) => {
                writeln!(f, "error: unterminated string")?;
                writeln!(f, "  |")?;
                writeln!(f, "  | string starting at position {} is not terminated", pos)
            }
            LexerError::UnexpectedEof => {
                writeln!(f, "error: unexpected end of file")?;
                writeln!(f, "  |")?;
                writeln!(f, "  | the file ends unexpectedly")
            }
        }
    }
}

impl fmt::Display for ParserError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ParserError::UnexpectedToken { expected, found, span } => {
                let file = span.file.as_deref().unwrap_or("unknown");
                writeln!(f, "error: unexpected token")?;
                writeln!(f, " --> {}:{}:{}", file, span.line, span.column + 1)?;
                writeln!(f, "  |")?;
                writeln!(f, "{} | {}", span.line, get_line_content(file, span.line).unwrap_or_default())?;
                writeln!(f, "  | {}{} expected one of {:?}, found '{}'", 
                    " ".repeat(span.column),
                    "^".repeat(found.len()),
                    expected,
                    found
                )
            }
            ParserError::MissingToken { expected, after, span } => {
                let file = span.file.as_deref().unwrap_or("unknown");
                writeln!(f, "error: missing token")?;
                writeln!(f, " --> {}:{}:{}", file, span.line, span.column + 1)?;
                writeln!(f, "  |")?;
                writeln!(f, "{} | {}", span.line, get_line_content(file, span.line).unwrap_or_default())?;
                writeln!(f, "  | {}{} missing '{}' after '{}'",
                    " ".repeat(span.column),
                    "^",
                    expected,
                    after
                )
            }
        }
    }
}

impl std::error::Error for LexerError {}
impl std::error::Error for ParserError {}

// Helper function to get the content of a specific line from a file
fn get_line_content(file: &str, line_number: usize) -> std::io::Result<String> {
    use std::io::{BufRead, BufReader};
    use std::fs::File;

    let file = File::open(file)?;
    let reader = BufReader::new(file);
    
    for (index, line) in reader.lines().enumerate() {
        if index + 1 == line_number {
            return Ok(line?);
        }
    }
    Ok(String::new())
} 