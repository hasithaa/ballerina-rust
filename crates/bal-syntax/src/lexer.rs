//! Lexer implementation for Ballerina

use logos::Logos;

#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub line: usize,    // 1-based line number
    pub column: usize,  // 0-based column
}

#[derive(Debug, Clone, PartialEq)]
pub struct TokenInfo {
    pub kind: Token,
    pub text: String,
    pub span: Span,
}

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(skip r"[ \t\n\f]+")]  // Ignore whitespace
pub enum Token {
    // Keywords
    #[token("import")]
    Import,
    
    #[token("public")]
    Public,
    
    #[token("function")]
    Function,
    
    #[token("returns")]
    Returns,
    
    #[token("int")]
    Int,
    
    #[token("boolean")]
    Boolean,
    
    #[token("if")]
    If,
    
    #[token("else")]
    Else,
    
    #[token("while")]
    While,
    
    #[token("break")]
    Break,
    
    #[token("continue")]
    Continue,
    
    #[token("return")]
    Return,
    
    #[token("true")]
    True,
    
    #[token("false")]
    False,
    
    // Operators
    #[token("==")]
    Eq,
    
    #[token("!=")]
    NotEq,
    
    #[token("<")]
    Lt,
    
    #[token("<=")]
    LtEq,
    
    #[token(">")]
    Gt,
    
    #[token(">=")]
    GtEq,
    
    #[token("+")]
    Plus,
    
    #[token("-")]
    Minus,
    
    #[token("*")]
    Star,
    
    #[token("/")]
    Slash,
    
    #[token("%")]
    Percent,
    
    #[token("=")]
    Assign,
    
    // Delimiters
    #[token("(")]
    LParen,
    
    #[token(")")]
    RParen,
    
    #[token("{")]
    LBrace,
    
    #[token("}")]
    RBrace,
    
    #[token(",")]
    Comma,
    
    #[token(":")]
    Colon,
    
    #[token(";")]
    Semicolon,
    
    // Identifiers and literals
    #[regex("[A-Za-z][A-Za-z0-9_]*")]
    Identifier,
    
    #[regex("0|[1-9][0-9]*")]
    IntegerLiteral,

    // Comments
    #[regex("//[^\n]*")]
    LineComment,
}

pub struct Lexer<'a> {
    inner: logos::Lexer<'a, Token>,
    source: &'a str,
    line_starts: Vec<usize>, // Cache line start positions
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        // Calculate line starts once during initialization
        let line_starts: Vec<_> = std::iter::once(0)
            .chain(input.match_indices('\n')
                .map(|(i, _)| i + 1))
            .collect();
        
        Self {
            inner: Token::lexer(input),
            source: input,
            line_starts,
        }
    }

    fn get_position(&self, offset: usize) -> (usize, usize) {
        match self.line_starts.binary_search(&offset) {
            Ok(line) => (line + 1, 0),
            Err(line) => (line, offset - self.line_starts[line - 1]),
        }
    }

    pub fn next_token(&mut self) -> Option<TokenInfo> {
        self.inner.next().map(|token| {
            let range = self.inner.span();
            let (line, column) = self.get_position(range.start);
            TokenInfo {
                kind: token.unwrap(),
                text: self.source[range.clone()].to_string(),
                span: Span {
                    start: range.start,
                    end: range.end,
                    line,
                    column,
                },
            }
        })
    }
} 