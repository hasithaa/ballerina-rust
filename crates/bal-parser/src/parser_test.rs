#[cfg(test)]
mod tests {
    use crate::Parser;
    use bal_syntax::SyntaxKind;
    use bal_syntax::error::Span;

    fn parse(input: &str) -> String {
        let tokens = tokenize(input);
        let parser = Parser::new(Some("test.bal".to_string()), tokens);
        let node = parser.parse();
        format!("{:#?}", node)
    }

    fn tokenize(input: &str) -> Vec<(SyntaxKind, String, Span)> {
        use bal_syntax::lexer::{Lexer, Token};
        
        let mut lexer = Lexer::new(input);
        let mut tokens = Vec::new();
        
        while let Some(result) = lexer.next_token() {
            if let Ok(token_info) = result {
                // Skip whitespace and newlines
                match token_info.kind {
                    Token::Import => tokens.push((SyntaxKind::IMPORT_KW, token_info.text, token_info.span)),
                    Token::Public => tokens.push((SyntaxKind::PUBLIC_KW, token_info.text, token_info.span)),
                    Token::Function => tokens.push((SyntaxKind::FUNCTION_KW, token_info.text, token_info.span)),
                    Token::Returns => tokens.push((SyntaxKind::RETURNS_KW, token_info.text, token_info.span)),
                    Token::Int => tokens.push((SyntaxKind::INT_KW, token_info.text, token_info.span)),
                    Token::Boolean => tokens.push((SyntaxKind::BOOLEAN_KW, token_info.text, token_info.span)),
                    Token::Identifier => tokens.push((SyntaxKind::IDENTIFIER, token_info.text, token_info.span)),
                    Token::LParen => tokens.push((SyntaxKind::L_PAREN, token_info.text, token_info.span)),
                    Token::RParen => tokens.push((SyntaxKind::R_PAREN, token_info.text, token_info.span)),
                    Token::LBrace => tokens.push((SyntaxKind::L_BRACE, token_info.text, token_info.span)),
                    Token::RBrace => tokens.push((SyntaxKind::R_BRACE, token_info.text, token_info.span)),
                    Token::Comma => tokens.push((SyntaxKind::COMMA, token_info.text, token_info.span)),
                    Token::Semicolon => tokens.push((SyntaxKind::SEMICOLON, token_info.text, token_info.span)),
                    Token::Slash => tokens.push((SyntaxKind::SLASH, token_info.text, token_info.span)),
                    Token::LineComment => tokens.push((SyntaxKind::COMMENT, token_info.text, token_info.span)),
                    Token::Newline => continue, // Skip newlines
                    _ => continue, // Skip other tokens
                }
            }
        }
        tokens
    }

    #[test]
    fn test_empty_file() {
        let input = "";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_import() {
        let input = "import foo/bar;";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_simple_function() {
        let input = "function add(int a, int b) returns int { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_public_function() {
        let input = "public function greet() { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_function_with_multiple_params() {
        let input = "function calc(int x, boolean flag, int y) { }";
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_complete_module() {
        let input = r#"
            import ballerina/io;
            
            public function add(int a, int b) returns int {
            }
            
            function helper(boolean flag) {
            }
        "#;
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_error_recovery_missing_semicolon() {
        let input = "import foo/bar";  // Missing semicolon
        insta::assert_snapshot!(parse(input));
    }

    #[test]
    fn test_error_recovery_missing_param_type() {
        let input = "function bad(a, int b) { }";  // Missing type for first param
        insta::assert_snapshot!(parse(input));
    }
} 