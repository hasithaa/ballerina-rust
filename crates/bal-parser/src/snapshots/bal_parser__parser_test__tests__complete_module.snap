---
source: crates/bal-parser/src/parser_test.rs
assertion_line: 88
expression: parse(input)
---
Ok(
    GreenNode {
        kind: SyntaxKind(
            39,
        ),
        text_len: 88,
        n_children: 3,
    },
)
